platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemSpec.java

```diff
@@ -18,87 +18,137 @@
 
 import org.gradle.api.Incubating;
 
-import javax.annotation.Nullable;
-
 /**
- * {@link Problem} instance builder that is not capable of creating a new instances.
- *
- * An example of how to use the builder:
- * <pre>{@code
- *  <problemReporter>.report(configurator -> configurator
- *          .label("test problem")
- *          .category("category", "subcategory")
- *          .severity(Severity.ERROR)
- *          .details("this is a test")
- *  }</pre>
+ * Provides options to configure problems.
+ * <p>
  *
+ * @see ProblemReporter
  * @since 8.6
  */
 @Incubating
 public interface ProblemSpec {
 
     /**
      * Declares a short message for this problem.
-     * @param label the short message
-     * @param args the arguments for formatting the label with {@link String#format(String, Object...)}
+     * <p>
+     * The label is the main, human-readable representation of the problem.
+     * It is a mandatory property to configure when emitting a problem with {@link ProblemReporter}.
      *
-     * @return the builder for the next required property
+     * @param label the short message
+     * @return this
      * @since 8.6
      */
-    ProblemSpec label(String label, Object... args);
```

Comment:
🤔 I think here we should use a typical `Logger`-like API:
 - A `label(String label)` method for messages not requiring any formatting
 - A `label(String label, Objects... args)` method for messages **requiring** formatting.

I think this is the best solution, which would solve both of our concerns about preventing an Exception from `String.format`, but at the same time, maximizing DX.

Reply:
I don't think that baking in `String.format` contract for a string is a good idea. The requirements for the format string are huge. I agree, it's less of a problem when it's just an overload, but still I don't think it's justified.

I would argue that the DX is actually better if the user decides to call `String.format` themselves. For one, IDEA even provides static analysis that makes sure your format string matches the number of arguments that is provided. By making the `format` call hidden inside `label` we take away these insights from the user.

Moreover, with time, there are more and more ways to interpolate a string. Kotlin has interpolation from the beginning, and it also provides functions like `buildString`. Even [Java introduces interpolation](https://openjdk.org/jeps/430), hinting that `String.format` is not always a perfect fit.

The loggers have such an API primarily for performance reasons, which I don't believe is applicable for Problems API. Also, loggers are [known to come up with their own formats](https://www.slf4j.org/faq.html#logging_performance), which dilutes the idea of universal expectation, and forces the users of the Problems API to read the javadocs (nobody reads javadoc ;)).

In other words, I think our API should not be biased towards `String.format`.

Reply:
I think at this point the logger format (especially after Java 8) is a bit of a moot point, and it is

Reply:
I've mentioned to @donat that I won't die on this hill.

I think, at this point, the logger format (especially after Java 8) is a convention rather than a limitation. It is indeed true that packing `String.format(...)`'s semantics is a heavy one; though as being used often by the developers, it's a well-known one. 

> For one, IDEA even provides static analysis that makes sure your format string matches the number of arguments that is provided.

True!

> Moreover, with time, there are more and more ways to interpolate a string. Kotlin has interpolation from the beginning, and it also provides functions like buildString. Even [Java introduces interpolation](https://openjdk.org/jeps/430), hinting that String.format is not always a perfect fit.

This is a bit optimistic view: how many of our plugin developers use Kotlin, or JDK21 (though the LTS situation might be helpful to tip avaw from JDK11, I'm not holding my breath)

---

In general, I agree with the points presented here, and let's go with the template-less solution!

Reply:
I agree that it is much simpler API wise to leave that up to the users of the API how they format their string. The only reason to add such an API would be performance, since it might be quite expensive to create a lot of dynamic strings. On the other hand,  I am not sure we could avoid the creation of the `String`s just because we offer such an API.